### PR TITLE
Add documentation for fatalError cases in Foundation

### LINKED ISSUE #118: Indicate when a force unwrap/fatalerror is "guarenteed" not to fail
My understanding is that a lot of the force unwraps/fatalerrors in Needle are expected to never crash because the code generator will ensure that they always succeed. 

It would be useful for code reviewers without a lot of context, or for users who might be debugging crashes in Needle, to know which code is guaranteed, and which code is potentially dangerous. 

This could take the form of new method of force unwrapping, for example:
```
extension Optional {
     var guarenteedUnwrap: Wrapped {
          return self!
     }
}
```
Or just be a commitment in the Needle Code Standard (if/when they're written) to only force unwrap when the result is guaranteed to succeed.

### PR DESCRIPTION
**Pull request by neakor**
Also did minor formatting for comments.

Fixes #118

### CODE REVIEW

File: Foundation/Sources/NeedleFoundation/Pluginized/PluginizedComponent.swift [hunk 1]
```diff
@@ -124,12 +127,17 @@ open class PluginizedComponent<DependencyType, PluginExtensionType, NonCoreCompo
...
         } else {
+            // This case should never occur with properly generated Needle code.
+            // Needle's official generator should guarantee the correctness.
             fatalError("Plugin extension provider factory for \(self) returned incorrect type. Should be of type \(String(describing: PluginExtensionType.self)). Actual type is \(String(describing: provider))")
         }
     }
 
     deinit {
         guard let lifecycleObserverDisposable = lifecycleObserverDisposable else {
+            // This occurs with inproper usages of a pluginized component. It
```
Comment (rudro) [inline Foundation/Sources/NeedleFoundation/Pluginized/PluginizedComponent.swift]: `inproper` -> `improper`

File: Foundation/Sources/NeedleFoundation/Pluginized/PluginizedComponent.swift [hunk 2]
```diff
@@ -58,6 +58,9 @@ open class PluginizedComponent<DependencyType, PluginExtensionType, NonCoreCompo
     /// directly access this property.
     public var nonCoreComponent: AnyObject {
         guard let value = releasableNonCoreComponent else {
+            // This case should not occur if the pluginized component is properly
+            // paried with a consumer. This only occurs if the `nonCoreComponent`
```
Comment (rudro) [inline Foundation/Sources/NeedleFoundation/Pluginized/PluginizedComponent.swift]: `paried` -> `paired`

File: Foundation/Sources/NeedleFoundation/Pluginized/PluginizedComponent.swift (hunk 1 above)
Comment (neakor) [reply Foundation/Sources/NeedleFoundation/Pluginized/PluginizedComponent.swift]: Spelling is hard. Done.

File: Foundation/Sources/NeedleFoundation/Pluginized/PluginizedComponent.swift (hunk 2 above)
Comment (neakor) [reply Foundation/Sources/NeedleFoundation/Pluginized/PluginizedComponent.swift]: Done.

**Pull request merged**